### PR TITLE
171094550 save template to s3

### PR DESCRIPTION
**Pull request by knowuh**
This is a big PR, but it's isolated.

The purpose is to add an authoring page which saves, loads, and verifies experiments templates. It uses S3 bucket access brokered by the Token Service.

[#171656112]
https://www.pivotaltracker.com/story/show/171656112

`src/authoring-app/utils/s3-resource-helper.ts` is an attempt to streamline S3 access, and the plan is to maybe move it to the token-service repo maybe with some additional tests.

### CODE REVIEW

File: src/authoring-app/components/resource-listing.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  disabled: boolean;
+  selectFn?: (resource: S3Resource) => void;
+}
+
+const ResourceList = (opts: IResourceListOpts) => {
+  const { resource, resources, selectFn, disabled} = opts;
+  const resElems = resources.map( (r:S3Resource) => {
+    const selectHandler = disabled
+      ? (e: React.MouseEvent<HTMLElement>) => null
+      : (e: React.MouseEvent<HTMLElement>) => { selectFn && selectFn(r); };
```
Comment (dougmartin) [inline src/authoring-app/components/resource-listing.tsx]: Do you need `e` or the type annotations here?

File: src/authoring-app/components/resource-selector.tsx [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+          Description:<br/><input type="text" ref={this.resourceDescriptionFieldRef} />
+        </p>
+        <p>
+          <Button
+            label="Create Resource"
+            data-cy="create-resource"
+            onClick={this.handleCreateResource}
+          />
+          <Button
+            label="Select Resiyrce"
```
Comment (dougmartin) [inline src/authoring-app/components/resource-selector.tsx]: Resiyrce => Resource

File: src/authoring-app/hooks/use-s3.tsx [hunk 3]
```diff
@@ -0,0 +1,227 @@
+import { useState } from "react";
+import { S3ResourceHelper, IS3ResourceHelperOpts} from "../utils/s3-resource-helper";
+import { S3Resource } from "@concord-consortium/token-service";
+import defaultExperiment from "../data/default-experiment.json";
+
+let helper:S3ResourceHelper = null as unknown as S3ResourceHelper;
```
Comment (dougmartin) [inline src/authoring-app/hooks/use-s3.tsx]: You can also type this as `helper: S3ResourceHelper | null` which would give you better type checking on references to `helper`.

File: src/authoring-app/hooks/use-s3.tsx [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+  const stageContentFn = (jsObject: object) => {
+    const json =JSON.stringify(jsObject, null, 2);
+    if(resourceContent !== json) {
+      setDirty(true);
+      setResourceContent(json);
+    }
+  };
+
+  const saveFn = async (newContent:string, newName:string, newDescription:string) => {
+    setStatus(S3Status.SavePending);
```
Comment (dougmartin) [inline src/authoring-app/hooks/use-s3.tsx]: Should this be moved inside the `if(s3Resource) {...`

File: src/authoring-app/hooks/use-validating-editor.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  }
+}
+
+/*
+* Ensure that JSON is parsable.
+* Returns array of errors. Empty array if clean.
+*/
+const lintJson = (json: string) => {
+	try {
+		JSON.parse(json);
```
Comment (dougmartin) [inline src/authoring-app/hooks/use-validating-editor.ts]: Looks like some space/tab formatting error here.

File: src/authoring-app/hooks/use-validating-editor.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+};
+
+// State magement for content, dirty, and invalid JSON in the editor:
+export const UseValidatingEditor = (
+  initialValue: string,
+  validationF: (value:string) => IErrorNotation[] = schemaValidation) => {
+
+    const [originalValue, setOriginalValue] = useState(initialValue);
+    const [editorValue, setEditorValue] = useState(initialValue);
+    const [errors, setErrors] = useState([] as IErrorNotation[]);
```
Comment (dougmartin) [inline src/authoring-app/hooks/use-validating-editor.ts]: For better typing you can use `useState<IErrorNotation[]>([])`

File: src/authoring-app/utils/getS3Config.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import { IS3ResourceHelperOpts} from "./s3-resource-helper";
+import { ResourceTool } from "@concord-consortium/token-service";
+import {getURLParam} from "../../shared/utils/get-url-param";
+
+export const PORTAL_URL_PARAM = "portalUrl";
+const devPortalUrl = "app.portal.docker";
```
Comment (dougmartin) [inline src/authoring-app/utils/getS3Config.tsx]: Mine is `app.rigse.docker` - is yours `app.portal.docker`?

**Comment by knowuh** [comment]
Thank you @dougmartin for the great suggestions.

Please have a peep at 6f289df  which should address your comments.

**Pull request merged**